### PR TITLE
support enable specify repeat connection to same server

### PR DESCRIPTION
**Pull request by changchengx**
## What
use new server info format ```<server-ip>:<port>[:<count>]``` to support repeat connect with same server.

## Why ?
process argument number is limited, e.g. getconf ARG_MAX. The current io_demo process can't accept too long arguments.

## How ?
support [server_ip:port:repeat_count] to specify the number for repeating connection with same server.

### CODE REVIEW

**Comment by yosefe** [comment]
@changchengx maybe instead of long command line better provide a file name with the list of servers?

**Comment by changchengx** [comment]
> @changchengx maybe instead of long command line better provide a file name with the list of servers?
>
@yosefe
1) Currently, there're no lots of servers infrastructure in the lab.
2) While issuing burst connections from client to server, two server hosts(client role & server role) are enough until all the QPs are used out. It doesn't need to enumerate too many **different** server info in the command line. **Most of the enumerated server info are same** and this PR uses ```-x``` to indicate how many times the server info should be repeated.
3) If we really need to enumerate **tens thousands of different server info** in the command line, we can ```provide a file name with the list of servers``` because the process's arguments length are limited by linux system(e.g. ```getconf ARG_MAX``` ).
Could we use this PR solution to solve the current problem wrote in the **Why?** section? I haven't imaged the scenario the it needs tens thousands of different server info and mock the scenario in the lab infrastructure.

**Comment by yosefe** [comment]
@changchengx so the problem this PR is trying to solve is connecting to same server multiple times without passing it in the command line multiple times.
WDYT to support a syntax of `<server-ip>:<port>[:<count>]`, for example 1.1.1.1:3245:100, which would mean to create 100 connections to 1.1.1.1 port 3245. if not provided, the count would be 1.

**Comment by changchengx** [comment]
@yosefe Agree with the ```syntax of <server-ip>:<port>[:<count>]``` solution. I'll follow this idea to implement the code.

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -2664,6 +2669,38 @@ static void adjust_opts(options_t *test_opts) {
         test_opts->chunk_size = std::min(test_opts->chunk_size,
                                          test_opts->max_data_size);
     }
+
+    if (servers.empty()) {
+        return;
+    }
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2675]: why needed?

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -2664,6 +2669,38 @@ static void adjust_opts(options_t *test_opts) {
...
+
+    for (auto server = servers.cbegin(); server != servers.cend(); ++server) {
+        repeat_separator = strchr(*server, ':');
+        if (repeat_separator == NULL) {
+            test_opts->servers.push_back(std::string(*server));
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2680]: can remove std::string it will use the constructor automatically

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -2664,6 +2669,38 @@ static void adjust_opts(options_t *test_opts) {
         test_opts->chunk_size = std::min(test_opts->chunk_size,
                                          test_opts->max_data_size);
     }
+
+    if (servers.empty()) {
+        return;
+    }
+
+    for (auto server = servers.cbegin(); server != servers.cend(); ++server) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2677]: can we use for (auto server: servers)

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -2664,6 +2669,38 @@ static void adjust_opts(options_t *test_opts) {
...
+        repeat_separator = strchr(*server, ':');
+        if (repeat_separator == NULL) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2679]: isn't this port_separator?

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -2916,10 +2955,10 @@ static int parse_args(int argc, char **argv, options_t *test_opts)
     }
 
     while (optind < argc) {
-        test_opts->servers.push_back(argv[optind++]);
+        servers.push_back(argv[optind++]);
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2958]: maybe instead of using local server array, we can call something like add_servers(argv[optind++]) for each argument. and add each server including repeat count if needed

File: test/apps/iodemo/io_demo.cc [hunk 6]
```diff
@@ -2664,6 +2669,38 @@ static void adjust_opts(options_t *test_opts) {
...
+        for (auto i = 0; i < repeat_count; i++) {
+            test_opts->servers.push_back(server_repeated);
+
+            if (test_opts->src_addrs.empty()) {
+                continue;
+            }
+
+            /* align with round robin method to select the src address */
+            test_opts->src_addrs.push_back(test_opts->src_addrs.at(
+                        std::distance(servers.cbegin(), server)));
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2701]: can't it exceed the size of src_addrs array?

Comment (changchengx) [reply test/apps/iodemo/io_demo.cc:2701]: I'll fix the wrong logic here.

**Comment by changchengx** [comment]
I'll refine this PR this today or next weekend until passing all the CI test cases.

File: test/apps/iodemo/io_demo.cc [hunk 7]
```diff
@@ -2916,10 +2953,19 @@ static int parse_args(int argc, char **argv, options_t *test_opts)
...
 
-    adjust_opts(test_opts);
+    if (!test_opts->src_addrs.empty() &&
+        (test_opts->src_addrs.size() != 1) &&
+        (test_opts->src_addrs.size() != servers.size())) {
+        std::cout << "In client size, source addr number should be 1 "
+                  << "OR equal server addr number" << std::endl;
+        return -1;
+    }
+
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2966]: why is this needed?

File: test/apps/iodemo/io_demo.cc (hunk 5 above)
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:2956]: can we round-robin the src addresses between all servers that we created (either directly or by repeat count)?
IMO it will make the code simpler and more generic - remove the dependency between test_opts->src_addrs.size() and servers.size()

File: test/apps/iodemo/io_demo.cc [hunk 8]
```diff
@@ -2664,6 +2670,35 @@ static void adjust_opts(options_t *test_opts) {
...
+        repeat_separator = strchr(port_separator + 1, ':');
+        if (repeat_separator == NULL) {
+            test_opts->servers.push_back(servers[idx]);
+            continue;
+        }
+
+        server_repeated =
+            std::string(servers[idx], repeat_separator - servers[idx]);
+        repeat_count    = atoi(repeat_separator + 1);
+        for (int i = 0; i < repeat_count; i++) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2690]: To solve Coverity error, probably need to set upper bound for repeat_count, for example
```cpp
#define MAX_SERVER_REPEAT_COUNT 100000ul
...
int parse_server(const char *server_arg) {
    ...
    repeat_count = atoi(repeat_separator + 1);
    if (repeat_count > MAX_SERVER_REPEAT_COUNT) {
        std::cout << "Exceeded maximal server repeat count (" << MAX_SERVER_REPEAT_COUNT << ")" << std::endl;
        return -1;
    }
```

File: test/apps/iodemo/io_demo.cc (hunk 5 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc:2956]: >
>can we round-robin the src addresses between all servers that we created (either directly or by repeat count)?
>
I didn't understand it well.
1. ```src_addrs``` is used one by one to connect with the server.
2. suppose there're two source addresses ```src_ip0, src_ip1``` and there're two servers ```server0:port0, server1:port1:2```.
    1) The current logic is to use ```src_ip0``` to connect with ```server0:port0``` and use ```src_ip1``` to connect with ```server1:port1``` for ```2``` times.
    2) Do you mean that it should ```src_ip0``` to connect with ```server0:port0``` and use ```src_ip1``` to connect with ```server1:port1``` for the first time, then use ```src_ip0``` to connect with ```server1:port1``` for the second time?

File: test/apps/iodemo/io_demo.cc (hunk 7 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc:2966]: It requires the src_addrs are used to connect with the server one by one
1) ```empty``` case: use default destination reachable src ip to connect with server.
2) ```1``` case: the only given src ip could connect with all servers.
3) ```src_addrs.size() = servers.size()``` case: the src addresses are used one by one to connect with the servers.

File: test/apps/iodemo/io_demo.cc (hunk 5 above)
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:2956]: in case (2), there are actually 3 servers: server0:port0, server1:port1, server1:port1
yes, i would suggest in this case we use src_addrs in the following order: 
server0:port0 - src_ip0
server1:port1 - src_ip1
server1:port1 - src_ip0

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:2956]: it's up to the user to specify right combination of server and src_addr

Comment (changchengx) [reply test/apps/iodemo/io_demo.cc:2956]: Got it. I'll simplify the code here and update this PR implementation on weekend time.

File: test/apps/iodemo/io_demo.cc (hunk 1 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done

File: test/apps/iodemo/io_demo.cc (hunk 4 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done

File: test/apps/iodemo/io_demo.cc (hunk 6 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done. Logic has been removed after simplifying the design

File: test/apps/iodemo/io_demo.cc (hunk 7 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Removed the check here after simplifying the design

File: test/apps/iodemo/io_demo.cc (hunk 8 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done. I also check whether the ```repeat_count``` is ```0``` here to 1) check wrong input 2) ```0``` should not be used by client.

File: test/apps/iodemo/io_demo.cc (hunk 5 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: The code is simplified a lot after following this idea to make the design.

**Comment by changchengx** [comment]
The failure in CI test case ```Tests BlueField on worker 0``` is not related with this PR. The error log is below:
```##[error]The job running on agent swx-rain03-bf1-01 ran longer than the maximum time of 360 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134```

File: test/apps/iodemo/io_demo.cc [hunk 9]
```diff
@@ -2682,6 +2683,40 @@ static int parse_window_size(const char *optarg, long &window_size,
...
+    if (repeat_separator == NULL) {
+        servers.push_back(server);
+        return 0;
+    }
+
+    server_repeated = std::string(server, repeat_separator - server);
+    repeat_count    = atoi(repeat_separator + 1);
+    if ((repeat_count == 0) || (repeat_count > MAX_SERVER_REPEAT_COUNT)) {
+        std::cout << "repeat connection number should be in range "
+                  << "[1, " << MAX_SERVER_REPEAT_COUNT << "]" << std::endl;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:2709]:
```cpp
    std::cout << "Server repeat_count should be in the range "
              << "[1.. " << MAX_SERVER_REPEAT_COUNT << "]" << std::endl;
```

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:2709]: @changchengx unresolved

Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: @yosefe Done

**Review by yosefe — DISMISSED** [review]
pls squash

**Review by yosefe — APPROVED** [review]
